File: civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

/**
 *
 * @author nacho
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSopresa;
    
    private void init(){
        sorpresas = new ArrayList <>();
        cartasEspeciales = new ArrayList <>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean d){
        this.init();
        this.debug = d;
        if(d)
            Diario.getInstance().ocurreEvento("Debug del mazo sorpresa activo");
    }
    
    MazoSorpresas(){
        this.init();
        this.debug = false;
    }
    
    void alMazo(Sorpresa s){
        if (!this.barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente(){ 
        if(!this.barajada || usadas==sorpresas.size() && !debug){
            Collections.shuffle(sorpresas);
            this.usadas = 0;
            this.barajada = true;
        }
        
        this.usadas++;
        this.sorpresas.add(sorpresas.get(0));
        this.ultimaSopresa = this.sorpresas.get(0);
        this.sorpresas.remove(0);
        
        return this.ultimaSopresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        boolean encontrado = false;
        for( int i=0; i< this.sorpresas.size() && !encontrado; i++)
        {
            if (sorpresa == this.sorpresas.get(i))
            {
                this.sorpresas.remove(i);
                this.cartasEspeciales.add(sorpresa);
                Diario.getInstance().ocurreEvento("Se ha inhabilitado la carta especial");
                encontrado = true;
            }
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        boolean encontrado = false;
        for (int i=0; i<this.sorpresas.size() && !encontrado; i++)
        {
            if (sorpresa == this.sorpresas.get(i))
            {
                this.sorpresas.remove(i);
                this.sorpresas.add(sorpresa);
                Diario.getInstance().ocurreEvento("Se ha movido la sorpresa al final del mazo");
                encontrado = true;
            }
        }
    }
}

File: Ruby/mazo_sorpresas.rb
# encoding: utf-8
require_relative "diario.rb"
  require_relative "sorpresa.rb"
module Civitas
  

class MazoSorpresas
  def initialize
    @barajada = false
    @usadas = 0
    @sorpresas = Array.new
    @cartasEspeciales = Array.new
    @debug = false
  end
  
  public
  attr_reader :sorpresas
  def MazoSorpresas(modo)
    initialize
    debug = modo
    if(debug)
      Diario.instance.ocurre_evento("El modo debug está activado")
    end
  end
  
 
  
  
  
  def alMazo(s)
    if(!barajada)
      @sorpresas.push(s)
    end
  end
  
  def siguiente
    if(!barajada||@usadas==@sorpresas.length)
      if(!debug)
        @sorpresas.shuffle
        @usadas=0
        barajada=true
      end
    end
    @usadas+=1
    @sorpresas.push(@sorpresas[0])
    @ultimaSorpresa = @sorpresas[0]
    @sorpresas.delete_at(0)
    return @ultimaSorpresa
  end
  
  def inhabilitarCartaEspecial(sorpresa)
    for i in(0..@sorpresas.length)
      if(@sorpresas[i]==sorpresa)
        @sorpresas.delete_at(i)
        @cartasEspeciales.push(sorpresa)
        Diario.instance.ocurre_evento("La carta especial ha sido inhabilitada")
      end
    end
  end
  
  def habilitarCartaEspecial(sorpresa)
    for i in(0..@cartasEspeciales.length)
      if(@cartasEspeciales[i]==sorpresa)
        @sorpresas.push(sorpresa)
        @cartasEspeciales.delete_at(i)
        Diario.instance.ocurre_evento("La carta especial ha sido habilitada")
      end
    end
  end
  attr_accessor:debug
  attr_accessor:barajada
end
end